enterprise/server/remote_execution/cgroup/cgroup.go

```diff
@@ -31,6 +32,105 @@ const (
+func Setup(ctx context.Context, path string, s *scpb.CgroupSettings, blockDevice *block_io.Device) error {
+	m, err := settingsMap(s, blockDevice)
+	if err != nil {
+		return err
+	}
+	for name, value := range m {
+		settingFilePath := filepath.Join(path, name)
+		if err := os.WriteFile(settingFilePath, []byte(value), 0); err != nil {
+			if os.IsNotExist(err) {
+				log.CtxWarningf(ctx, "Failed to set up cgroup file %q: %s", settingFilePath, err)
+			} else {
+				return fmt.Errorf("write %q: %w", settingFilePath, err)
+			}
+		}
+	}
+	return nil
+}
+
+func settingsMap(s *scpb.CgroupSettings, blockDevice *block_io.Device) (map[string]string, error) {
+	m := map[string]string{}
+	if s.CpuWeight != nil {
+		m["cpu.weight"] = fmt.Sprintf("%d", s.GetCpuWeight())
+	}
+	if s.CpuQuotaPeriodUsec != nil {
+		m["cpu.max"] = fmt.Sprintf("%d %s", s.GetCpuQuotaLimitUsec(), strconv.Itoa(int(s.GetCpuQuotaPeriodUsec())))
+	}
+	if s.CpuMaxBurstUsec != nil {
+		m["cpu.max.burst"] = fmt.Sprintf("%d", s.GetCpuMaxBurstUsec())
+	}
+	if s.CpuUclampMin != nil {
+		m["cpu.uclamp.min"] = fmtPercent(s.GetCpuUclampMin())
+	}
+	if s.CpuUclampMax != nil {
+		m["cpu.uclamp.max"] = fmtPercent(s.GetCpuUclampMax())
+	}
+	if s.MemoryThrottleLimitBytes != nil {
+		m["memory.high"] = strconv.Itoa(int(s.GetMemoryThrottleLimitBytes()))
+	}
+	if s.MemoryLimitBytes != nil {
+		m["memory.max"] = strconv.Itoa(int(s.GetMemoryLimitBytes()))
+	}
+	if s.MemorySoftGuaranteeBytes != nil {
+		m["memory.low"] = strconv.Itoa(int(s.GetMemorySoftGuaranteeBytes()))
+	}
+	if s.MemoryMinimumBytes != nil {
+		m["memory.min"] = strconv.Itoa(int(s.GetMemoryMinimumBytes()))
+	}
+	if s.SwapThrottleLimitBytes != nil {
+		m["memory.swap.high"] = strconv.Itoa(int(s.GetSwapThrottleLimitBytes()))
+	}
+	if s.SwapLimitBytes != nil {
+		m["memory.swap.max"] = strconv.Itoa(int(s.GetSwapLimitBytes()))
+	}
+	if blockDevice != nil {
+		if s.BlockIoLatencyTargetUsec != nil {
+			m["io.latency"] = fmt.Sprintf("%d:%d target=%d", blockDevice.Maj, blockDevice.Min, s.GetBlockIoLatencyTargetUsec())
+		}
+		if s.BlockIoWeight != nil {
+			m["io.weight"] = fmt.Sprintf("%d:%d %d", blockDevice.Maj, blockDevice.Min, clampWeight(s.GetBlockIoWeight()))
+		}
+		var limitFields []string
+		limits := s.GetBlockIoLimit()
+		if limits != nil {
+			if limits.Riops != nil {
+				limitFields = append(limitFields, fmt.Sprintf("riops=%d", limits.GetRiops()))
+			}
+			if limits.Wiops != nil {
+				limitFields = append(limitFields, fmt.Sprintf("wiops=%d", limits.GetWiops()))
+			}
+			if limits.Rbps != nil {
+				limitFields = append(limitFields, fmt.Sprintf("rbps=%d", limits.GetRbps()))
+			}
+			if limits.Wbps != nil {
+				limitFields = append(limitFields, fmt.Sprintf("wbps=%d", limits.GetWbps()))
+			}
+		}
+		if len(limitFields) > 0 {
+			m["io.max"] = fmt.Sprintf("%d:%d %s", blockDevice.Maj, blockDevice.Min, strings.Join(limitFields, " "))
+		}
+	}
+	return m, nil
+}
+
+func clampWeight(v int64) int64 {
```

Comment:
I'm thinking that it might be better to fail requests with invalid values instead of clamping them. That will make it faster to find bugs in the execution and scheduler servers.

Reply:
agreed, done